### ISSUE TITLE
Detect HTML5 foster-parenting violations (table content-model)

### ISSUE THREAD

**Issue by YusukeHirao** [markuplint/markuplint#3827]
## Problem

HTML5 parsing's foster-parenting rules (HTML LS §13.2.6.4.10 "in table" insertion mode and related) move forbidden child elements OUT of `<table>` at parse time. The author's source markup intent is lost from the resulting DOM, so DOM-walking rules like `permitted-contents` cannot see the violation.

## Affected bench fixtures (assertions slice)

- `tests/external/validator/tests/html/elements/table/model-input-child-novalid.html` — `<table><input></table>`
- `tests/external/validator/tests/html/elements/table/model-input-type-child-novalid.html` — `<table><input type=submit></table>`

Both stay `nu-only` after PR #3826 (assertions slice) because nu-validator parses with source-position awareness while markuplint walks the post-foster-parented DOM.

## Reproduction

```js
import { mlTest } from 'markuplint';
const r = await mlTest('<table><input></table>', { rules: { 'permitted-contents': true } });
// r.violations.length === 0  (expected: 1, input is a forbidden child of table)
```

The HTML parser's nodeList preserves source order but assigns `parentNode: null` to the foster-parented element, and `Document.walkOn('Element', ...)` hands rules a DOM where `<input>` is a sibling of `<table>`, not a child.

## Other foster-parenting cases worth covering

**Correction (2026-08-15):** foster parenting is *not* a table-wide effect. Per HTML LS §13.2.6.4, only three insertion modes ever foster-parent an unhandled token: "in table" (§13.2.6.4.9), "in table body" (§13.2.6.4.13), and "in row" (§13.2.6.4.14) — each one's "Anything else" entry either enables foster parenting directly or delegates to "in table"'s rules, which does. "In cell" (§13.2.6.4.15, active once inside `<td>`/`<th>`) delegates its "Anything else" entry to **"in body"** instead — ordinary insertion, no foster parenting. Verified directly against the parser: `<table><tr><input></tr></table>` puts `<input>` as a child of `<body>` (foster-parented), but `<table><tr><td><input></td></tr></table>` puts it as a child of `<td>` (normal, zero violations — this is valid markup, not a gap).

So the actual affected surface is: any non-table-construction token appearing directly inside `<table>`, `<tbody>`/`<thead>`/`<tfoot>`, or `<tr>` — not inside a `<td>`/`<th>` cell. Confirmed still affects:

- `<table>` containing any non-allowed child directly (`<div>`, `<p>`, etc.), not wrapped in a cell
- `<select>` containing `<table>` (foster-parents the table's content — same "in table" family)
- Any `in-table`/`in-table-body`/`in-row` insertion-mode token that the spec forbids

## Spec confirms this is a mandated Parse error, and parse5 is silent on it

HTML LS §13.2.6.4.9 ("in table" insertion mode), "A start tag whose tag name is 'input'": if the token has no `type` attribute, or `type` is not an ASCII case-insensitive match for `hidden`, act as "Anything else". The "Anything else" entry (verbatim): "**Parse error**. Enable foster parenting, process the token using the rules for the 'in body' insertion mode, and then disable foster parenting."

Probed parse5 directly (same technique used for #3943's parse5-silent-recovery rules — `parse5.parse(html, { onParseError })`): `<!doctype html><table><input></table>` emits **zero** parse error events. parse5 silently performs the foster-parenting tree correction without ever surfacing the spec-mandated parse error. This is the exact same shape of gap that #3943 already solved with 4 new dedicated rules (`no-unclosed-element-at-eof`, `no-content-after-body`, `no-stray-head-or-body-tag`, `meta-charset-position`) — **not** a parser-architecture problem requiring token-stream access or new AST fields.

## Possible approaches

1. ~~Token-stream rule~~ — unnecessary; superseded by approach 2 below now that it's confirmed sufficient.
2. **Source-position correlation (recommended)** — for every `<table>` element T with source range `[T.startOffset, T.closeTag?.endOffset ?? EOF]`, find any element E whose `startOffset` falls inside that range but which does **not** have T as an ancestor via `parentElement` (i.e. it was foster-parented out). This is generic — no `input`-specific logic, no new AST fields — and correctly ignores legitimately-nested content at any depth (e.g. `<table><td><table>...` — the outer T is still found as an ancestor by walking `parentElement`). `<script>`/`<style>`/`<template>` subtrees need the same RAWTEXT exclusion `no-stray-head-or-body-tag` already uses (they're processed under "in head" rules, not foster-parented).
3. ~~Parser annotation~~ — unnecessary for the same reason as (1); the existing `startOffset`/`parentElement` data on nodes already fully determines this without new fields.

## Open design question

Unlike #3943's rules (opt-in, `packages/@markuplint/rules/src/no-*`), this is arguably closer to the built-in `parseError` channel (`severity.parseError`) — it literally *is* a spec "Parse error" that parse5 just doesn't emit a code for, same category as parse5's own native codes. Two options, undecided:

- **New dedicated rule** (e.g. `no-foster-parented-content`), matching the #3943 precedent exactly.
- **Synthesize a parse-error code** in `@markuplint/html-parser` itself (something like `foster-parenting-content`), surfaced through the same `document.parseErrors` / `severity.parseError` channel as parse5's native codes, since the spec explicitly frames it as "Parse error" rather than a content-model rule concern.

The tradeoff: the `parseError` channel is more spec-faithful (this literally is one) and free for every rule config without an extra `rules` entry, but #3943 chose the dedicated-rule shape for a directly analogous case (also a parse5-silent, spec-mandated parse error) — for consistency, and because it lets each case get its own default severity / bench toggle / README independently. Needs a decision before implementation.

## Context

- Surfaced during PR #3826 (assertions slice) follow-up. The 2 fixtures above were intentionally deferred from that PR.
- bench-triage SKILL audit log does NOT cover this case; it's a real markuplint coverage gap, not nu over-detection.
- markuplint version: 5.0.0-rc.4 (v5 stable prep)

## Acceptance criteria

- The 2 fixtures above flip from `nu-only` to `match-error` in `tests/external/snapshots/diff/coverage.json`.
- A regression test covers `<table><input></table>` and at least one other foster-parenting case.
- The chosen approach is documented in `packages/@markuplint/parser-utils/SKILL.md` or a new dedicated SKILL so future contributors can extend it.
